CORE-13706 - Add 'force' flag to delete API

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Adds a `force` flag to the `DeleteShadowLinkRequest` message to permit users to delete a shadow link even if there are active shadow topics.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
cl: Implement force flag

When delete is called with a force flag, then attempt to fail over all
topics before link deletion.  This will hold up delete until all topics
have failed over.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto: Add force flag to delete

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin: Use force flag

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Add force option to delete link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -396,6 +396,8 @@ enum ShadowLinkState {
     SHADOW_LINK_STATE_ACTIVE = 1;
     // Shadow link was paused
     SHADOW_LINK_STATE_PAUSED = 2;
+    // Shadow link is in the process of being deleted
+    SHADOW_LINK_STATE_DELETING = 3;
```
tbh I'm not sure if I need this anymore....

bharathv replied:
ya I don't see why this is needed.

michael-redpanda replied:
yeah originally i was thinking of doing the failover as part of force delete but now tha tI'm not I think I can drop a lot of these changes...

bharathv commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -156,6 +157,18 @@ ss::future<errc> frontend::update_cluster_link_configuration(
     co_return co_await do_mutation(std::move(c), timeout);
 }
 
+ss::future<errc> frontend::update_shadow_link_status(
```
do we still need this API if we are using "force" boolean in DeleteShadowLink API?

bharathv commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -563,34 +563,7 @@ errc frontend::validator::validate_mutation(const cluster_link_cmd& cmd) const {
           if (!meta.has_value()) {
               return errc::does_not_exist;
           }
-          const auto& md = meta->get();
```
umm I was thinking we'd still need these checks if force = false?

michael-redpanda replied:
I was thinking we could depend upon the caller to know whether deletion was appropriate or not?

bharathv commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -102,6 +102,9 @@ message DeleteShadowLinkRequest {
             type: "redpanda.core.admin.ShadowLinkService/ShadowLink"
         }
     ];
+    // If true, then any active shadow topics will be failed over first before
+    // the link is deleted
```
In my mind force seems to address a different problem

Failover (i.e., opening the Kafka API) occurs regardless of whether force=true or force=false is used.

* When force=false (default), guardrails are in place to prevent the deletion of links that have active mirroring topics. This is designed to protect against accidental deletion by users.

* When force=true, the intent is explicit: proceed with deletion regardless of any active mirroring topics. In this case, the link is deleted, and the topics become active.

michael-redpanda replied:
that comment is out of date

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--94bd8816-4232-4679-821c-bd66cd52c81e-->
<!--TEST RESULTS-->
<!--5e2e4ed4-6b5b-496e-8661-5d3380a214f4-->
<!--TEST RESULTS-->
<!--14db79e5-0102-4fe1-b31b-ab4c4ec7fec5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73555">build#73555</a></summary>


| test_class                         | test_method                   | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                   |
|:-----------------------------------|:------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests      | test_replication_basic        | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}     | integration | https://buildkite.com/redpanda/redpanda/builds/73555#0199ab4f-dc63-4246-be68-12c612290262 | FLAKY         | 20/21    | upstream reliability is '99.51690821256038'. current run reliability is '95.23809523809523'. drift is 4.27881 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic             |
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73555#0199ab4f-dc62-4915-bc93-c8f838920961 | FLAKY         | 20/21    | upstream reliability is '98.4251968503937'. current run reliability is '95.23809523809523'. drift is 3.1871 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogLimitMirrorMakerTests&test_method=test_mirror_maker_with_limits |
| RandomNodeOperationsTest           | test_node_operations          | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73555#0199ab4f-dc5d-4ba1-8d6c-392a7b56b2a7 | FLAKY         | 20/21    | upstream reliability is '97.15189873417721'. current run reliability is '95.23809523809523'. drift is 1.9138 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                    |
| TimeQueryTest                      | test_timequery_with_local_gc  | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73555#0199ab4f-dc62-4915-bc93-c8f838920961 | FLAKY         | 18/21    | upstream reliability is '92.13349225268176'. current run reliability is '85.71428571428571'. drift is 6.41921 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73631">build#73631</a></summary>


| test_class                                      | test_method                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                              |
|:------------------------------------------------|:---------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest                           | test_mount_inexistent      | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73631#0199b9fb-611e-4388-b9c4-ef0cd26bac72 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent |
| TLSVersionTestECDSA                             | test_change_version        | {"version": 0}   | integration | https://buildkite.com/redpanda/redpanda/builds/73631#0199b9fb-6122-4629-a60c-144610ac2791 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TLSVersionTestECDSA&test_method=test_change_version     |
| update_properties_invalid_describe_configs_test | return_wrong_resource_type |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/73631#0199b9dc-f0df-48d2-90c9-6b0d790656bf | FAIL          | 0/1      |                                                                                                                                                             |                                                                                                                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73653">build#73653</a></summary>


| test_class                          | test_method         | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:------------------------------------|:--------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TieredStorageTest                   | test_tiered_storage | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73653#0199bb61-867f-4740-b607-819cfd10a51a | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage              |
| WriteCachingFailureInjectionE2ETest | test_crash_all      | {"use_transactions": false}                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/73653#0199bb61-867d-4967-b315-b5eef44bc185 | FLAKY         | 18/21    | upstream reliability is '83.92116182572614'. current run reliability is '85.71428571428571'. drift is -1.79312 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

IoannisRP commented on src/v/cluster/cluster_link/table.cc:
```diff
@@ -526,4 +533,19 @@ cluster::cluster_link::errc table::update_cluster_link_configuration(
+  const id_t id,
+  const update_shadow_link_status_cmd& cmd,
+  model::revision_id revision) {
+    if (!_link_metadata.contains(id)) {
+        vlog(cluster::clusterlog.warn, "Link id {} not found", id);
+    }
```
all other paths return `return errc::does_not_exist;` in the "Link id {} not found" section. This continues after the warning.

michael-redpanda replied:
this code is being removed

IoannisRP commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -111,21 +111,45 @@ shadow_link_service_impl::create_shadow_link(
-  serde::pb::rpc::context, proto::admin::delete_shadow_link_request req) {
+  serde::pb::rpc::context ctx, proto::admin::delete_shadow_link_request req) {
     vlog(sllog.trace, "delete_shadow_link: {}", req);
+    auto redirect_node = redirect_to(model::controller_ntp);
+    if (redirect_node) {
+        vlog(
+          sllog.debug,
+          "Redirecting to leader of {}: {}",
+          model::controller_ntp,
+          *redirect_node);
+        co_return co_await _proxy_client
+          .make_client_for_node<proto::admin::shadow_link_service_client>(
+            *redirect_node)
+          .delete_shadow_link(ctx, std::move(req));
+    }
```
some of these functions have redirect and some don't. Should they all have it?

michael-redpanda replied:
done

IoannisRP commented:
there is a bit of mismatch of naming across "{cluster|shadow}_link" and "{mirror|shadow}_topic" across the different components. I wonder if it would be worth the effort to streamline naming...

michael-redpanda commented:
> there is a bit of mismatch of naming across "{cluster|shadow}_link" and "{mirror|shadow}_topic" across the different components. I wonder if it would be worth the effort to streamline naming...

There will be

michael-redpanda commented:
Force push:

* Removed `deleting` state and code surrounding that

michael-redpanda commented:
Force push:

* Rebased off of dev

bharathv commented on src/v/cluster_link/manager.cc:
```diff
@@ -368,13 +371,22 @@ ss::future<cl_result<void>> manager::delete_cluster_link(model::name_t name) {
-            "Failed to delete cluster link with name '{}'. There are "
-            "active/promoting shadow topics.",
-            name));
+    auto active_shadow_topics = std::ranges::any_of(
+      mirror_topic_states, is_active);
+
+    if (active_shadow_topics) {
+        if (!force_delete_link) {
+            co_return err_info(
+              errc::link_has_active_shadow_topics,
+              fmt::format(
+                "Failed to delete cluster link with name '{}'. There are "
+                "active/promoting shadow topics.",
+                name));
+        }
+        vlog(
+          cllog.info,
+          "Force deleting link \"{}\" with active shadow topics",
+          name);
```
wondering why the same validation does not exist in frontend.cc? (the latter is the one that runs under lock prior to replication, right?)

michael-redpanda replied:
In my opinion, the job of frontend is to ensure the table doesn't get into a weird state or that the command isn't operating on non-existant items in the table.  A delete command, at the moment, is just the name of the link to delete.  I could expand it to also include the force flag, but I don't feel it's necessary.  The manager acts as the gateway for the command to ensure the user isn't doing something they didn't mean to do.

bharathv replied:
I didn't fully follow, having a check in manager.cc is not tight enough, right, because it happens outside of the mutex? 

For example 

this check passes at t=0
topic_syncer adds a new active topic update at  t=1
frontend.cc applies this mutation t=2 (without any of these checks?)

I think frontend.cc::validate_mutation() is where all the requests converge and we have to validate under that lock?

michael-redpanda replied:
Fair point... i'll update the delete command... right now it only takes a link name, but I'll change it up to take a structure and stick the force flag in there

bharathv commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -102,6 +102,10 @@ message DeleteShadowLinkRequest {
             type: "redpanda.core.admin.ShadowLinkService/ShadowLink"
         }
     ];
+    // By default, DeleteShadowLink will fail if there are any active Shadow
+    // Topics.  Set this flag to 'true' if you wish to delete the link while
+    // there are active shadow topics.
```
Unclear what this means if the topic is in "failing_over / promoting". I feel this should preserve the original behavior i.e., if any topic is not failed_over/promoted/failed we reject the deletion request. wdyt?

michael-redpanda replied:
I feel the purpose to `force` here is to allow a user to say "forget all state, just delete the link" and the link gets deleted, regardless of what state the topics are in.

bharathv replied:
I see ok, not a blocker IMO, we can always change this condition depending on requirements. I'm fine with the current approach.

bharathv reviewed (commented):
couple of clarifications

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73631-->
<!--36aa29d3-bca6-429b-915a-caa96886aeb9-->
<!--title-->
## Retry command for [Build#73631](https://buildkite.com/redpanda/redpanda/builds/73631)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tls_version_test.py::TLSVersionTestECDSA.test_change_version@{"version":0}
```
<!--/body-->

michael-redpanda committed:
cl/model: Add delete_shadow_link_cmd

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Updated remove link commands

Remove link command used to only take a name as the 'key' in the
controller command.  Now it has been updated that the value takes a
structure with both the name and whether or not the delete was 'forced'.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Updated remove cluster link command to include the 'force' flag

bharathv commented on src/v/cluster_link/model/types.h:
```diff
@@ -909,6 +910,19 @@ struct aggregated_shadow_topic_report {
 };
 using report_result_t = std::expected<aggregated_shadow_topic_report, errc>;
+
+struct delete_shadow_link_cmd {
```
extend an envelope type?

bharathv reviewed (commented):
minor comment, lgtm otherwise

michael-redpanda commented:
Force push:

* Fixed upc ommit history

IoannisRP commented on src/v/cluster/cluster_link/tests/frontend_validation_test.cc:
```diff
@@ -80,12 +80,15 @@ class frontend_validation_test : public seastar_test {
-    ss::future<cluster::cluster_link::errc> delete_cluster_link(name_t m) {
-        cluster::cluster_link_remove_cmd cmd{std::move(m), 0};
+    ss::future<cluster::cluster_link::errc>
+    delete_cluster_link(name_t m, bool force) {
```
nit: default force to `false`?
simplifies changes and aligns with expectations

Same for `create_remove_command`

IoannisRP reviewed (approved):
looks good!

paulohtb6 commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -102,6 +102,10 @@ message DeleteShadowLinkRequest {
             type: "redpanda.core.admin.ShadowLinkService/ShadowLink"
         }
     ];
+    // By default, DeleteShadowLink will fail if there are any active Shadow
+    // Topics.  Set this flag to 'true' if you wish to delete the link while
+    // there are active shadow topics.
```
```suggestion
    // By default, DeleteShadowLink fails when active shadow
    // topics exists.  Set this flag to 'true' to delete the link even
    // when shadow topics are active.
```

nit for active voice. 

Note that I removed the capitalization on "Shadow Topic". We only captalize enterprise features. Objects should not be capitalized. This doesn't apply to `DeleteShadowLink` as it's an API object and we preserve their original form.

michael-redpanda replied:
Will do in follow up: https://redpandadata.atlassian.net/browse/CORE-13801
